src/include/cell_inline.h

```diff
@@ -1460,76 +1414,28 @@ __wt_cell_unpack_delta_int(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *page_
 }
 
 /*
- * __wt_cell_unpack_delta_leaf --
- *     Unpack a leaf delta cell into a structure.
+ * __wt_cell_unpack_delta_leaf_value --
+ *     Unpack a leaf delta value cell into a structure.
  */
 static WT_INLINE void
-__wt_cell_unpack_delta_leaf(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk,
-  WT_DELTA_CELL_LEAF *cell, WT_CELL_UNPACK_DELTA_LEAF *unpack)
+__wt_cell_unpack_delta_leaf_value(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk,
+  WT_CELL *value_cell, WT_CELL_UNPACK_DELTA_LEAF_KV *unpack)
 {
     WT_DECL_RET;
-    uint64_t v;
-    const uint8_t *p;
-
-    v = 0;
-
-    unpack->flags = cell->__chunk[0];
-    p = (uint8_t *)&cell->__chunk[1];
-
-    if (F_ISSET(unpack, WT_DELTA_LEAF_IS_DELETE)) {
-        ret = __wt_vunpack_uint(&p, 0, &v);
-        WT_ASSERT(session, ret == 0);
-        unpack->key_size = (uint32_t)v;
-        unpack->key = p;
-        unpack->__len = (uint32_t)WT_PTRDIFF(p + unpack->key_size, &cell->__chunk[0]);
-    } else {
-        WT_TIME_WINDOW_INIT(&unpack->tw);
-
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_START_TXN_ID)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.start_txn);
-            WT_ASSERT(session, ret == 0);
-        }
 
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_START_TS)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.start_ts);
-            WT_ASSERT(session, ret == 0);
-        }
-
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_START_DURABLE_TS)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.durable_start_ts);
-            WT_ASSERT(session, ret == 0);
-        }
-
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_STOP_TXN_ID)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.stop_txn);
-            WT_ASSERT(session, ret == 0);
-        }
-
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_STOP_TS)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.stop_ts);
-            WT_ASSERT(session, ret == 0);
-        }
+    /* Unpack the value. */
+    __wt_cell_unpack_kv(session, dsk, value_cell, &unpack->delta_value);
 
-        if (F_ISSET(unpack, WT_DELTA_LEAF_HAS_STOP_DURABLE_TS)) {
-            ret = __wt_vunpack_uint(&p, 0, &unpack->tw.durable_stop_ts);
-            WT_ASSERT(session, ret == 0);
-        }
-
-        ret = __wt_vunpack_uint(&p, 0, &v);
-        WT_ASSERT(session, ret == 0);
-        unpack->key_size = (uint32_t)v;
-        ret = __wt_vunpack_uint(&p, 0, &v);
-        WT_ASSERT(session, ret == 0);
-        unpack->value_size = (uint32_t)v;
-
-        unpack->key = p;
-        p += unpack->key_size;
-        unpack->value = p;
-        unpack->__len = (uint32_t)WT_PTRDIFF(p + unpack->value_size, &cell->__chunk[0]);
-
-        __cell_unpack_window_cleanup_delta(session, dsk, unpack);
+    if (unpack->delta_value.size > unpack->delta_leaf_value_data.size - 1) {
+        WT_IGNORE_RET(
+          __wt_buf_grow(session, &unpack->delta_leaf_value_data, unpack->delta_value.size - 1));
```

Comment:
Here's another implicit data copy. I don't understand why we need to allocate a buffer to delta_leaf_value_data

Reply:
@quchenhao We don't know how big the delta value will be until we've unpacked the cell.

Reply:
delta_leaf_value_data should only be a pointer. It doesn't need the size.

Reply:
Removed the buffer [here](https://github.com/wiredtiger/wiredtiger/pull/12223/commits/2ad349fc30605bf319d82b0052ddcf848d7ae1a3).